advisor/src/main/kotlin/advisors/VulnerableCode.kt

```diff
@@ -60,97 +54,69 @@ class VulnerableCode(
         val startTime = Instant.now()
 
-        val components = packages.map { it.purl }
-
         @Suppress("TooGenericExceptionCaught")
         return try {
-            val packagesToVulnerabilityRefs = mutableMapOf<String, VulnerableCodeService.Vulnerabilities>()
-
-            components.chunked(BULK_FETCH_SIZE).forEach { component ->
-                val requestResults = service.getPackageVulnerabilities(VulnerableCodeService.PackagesWrapper(component))
-                packagesToVulnerabilityRefs += requestResults
-            }
-
-            val packagesToVulnerabilities = fetchVulnerabilityDetails(packagesToVulnerabilityRefs, service)
-
-            val endTime = Instant.now()
-
-            packages.mapNotNull { pkg ->
-                packagesToVulnerabilities[pkg.id.toPurl()]?.let { vulnerabilities ->
-                    pkg to listOf(
-                        AdvisorResult(
-                            vulnerabilities,
-                            AdvisorDetails(providerName),
-                            AdvisorSummary(startTime, endTime)
-                        )
-                    )
+            mutableMapOf<Package, List<AdvisorResult>>().also {
+                packages.chunked(BULK_FETCH_SIZE).forEach { pkg ->
+                    it += loadVulnerabilities(pkg, startTime)
                 }
-            }.toMap()
+            }
         } catch (e: Exception) {
             createFailedResults(startTime, packages, e)
         }
     }
 
     /**
-     * Construct a map from package URLs to the vulnerabilities of this package based on a
-     * [map with vulnerability references][packagesToVulnerabilityRefs] using the specified [service]. This
-     * function looks up the details of the vulnerabilities the references point to.
+     * Load vulnerability information for the given [packages] and create a map with results per package using the
+     * [startTime].
      */
-    private suspend fun fetchVulnerabilityDetails(
-        packagesToVulnerabilityRefs: MutableMap<String, VulnerableCodeService.Vulnerabilities>,
-        service: VulnerableCodeService
-    ): Map<String, List<Vulnerability>> {
-        val vulnerabilityIds = packagesToVulnerabilityRefs.values.flatMap { it.unresolvedVulnerabilities }
-            .map { it.url }
-        val vulnerabilityDetails = resolveVulnerabilities(service, vulnerabilityIds)
-        return packagesToVulnerabilityRefs
-            .filterNot { it.value.unresolvedVulnerabilities.isEmpty() }
-            .mapValues { (_, refs) ->
-                refs.unresolvedVulnerabilities.mapNotNull { vulnerabilityDetails[it.url] }
+    private suspend fun loadVulnerabilities(
+        packages: List<Package>,
+        startTime: Instant
+    ): Map<Package, List<AdvisorResult>> {
+        val packageMap = packages.associateBy { it.purl }
+        val packageVulnerabilities =
+            service.getPackageVulnerabilities(VulnerableCodeService.PackagesWrapper(packageMap.keys.toList()))
```

Comment:
How about making `PackagesWrapper` take a `Collection` instead of `List` so we can avoid the conversion here? Should we do the same then also for other properties in classes of `VulnerableCodeService` for consistency?

Reply:
Also, maybe import `PackagesWrapper` so we can shorten the call to make it fit into a single line?

Reply:
I changed `PackageWrapper` to accept a collection. Regarding other properties, I think, in order to make the API convenient to use, we should distinguish between properties that need to be set by clients (like the collection in `PackageWrapper`), and the ones consumed by clients. For the latter it is more useful to use a more specific and powerful type, as this offers more possibilities to clients. The other properties fall into this category, so I would not change them.

Reply:
> we should distinguish between properties that need to be set by clients (like the collection in PackageWrapper), and the ones consumed by clients.

Good point, I agree. That basically matches the input / output principle in function signatures, where it's considered good practice to accept `Collection` in the input, but be specific (like `List`) in the output / return value.